packages/mui-utils/src/composeClasses/composeClasses.ts

```diff
@@ -1,30 +1,69 @@
+/* eslint no-restricted-syntax: 0, prefer-template: 0, no-unreachable-loop: 0
+   ---
+   These rules are preventing the performance optimizations below.
+ */
+
 export default function composeClasses<ClassKey extends string>(
   slots: Record<ClassKey, ReadonlyArray<string | false | undefined | null>>,
   getUtilityClass: (slot: string) => string,
   classes: Record<string, string> | undefined = undefined,
 ): Record<ClassKey, string> {
-  const output: Record<ClassKey, string> = {} as any;
+  const output = {} as Record<ClassKey, string>;
+
+  /* This is split in 2 loops for performance reasons, as it allows to avoid the `if (classes)`
+   * check in both of those cases. If `classes` is either `{}` or `undefined`, we consider it empty. */
+  if (isEmpty(classes)) {
+    for (const slotName in slots) {
+      if (Object.prototype.hasOwnProperty.call(slots, slotName)) {
+        const slotNames = slots[slotName];
 
-  Object.keys(slots).forEach(
-    // `Object.keys(slots)` can't be wider than `T` because we infer `T` from `slots`.
-    // @ts-expect-error https://github.com/microsoft/TypeScript/pull/12253#issuecomment-263132208
-    (slot: ClassKey) => {
-      output[slot] = slots[slot]
-        .reduce((acc, key) => {
-          if (key) {
-            const utilityClass = getUtilityClass(key);
-            if (utilityClass !== '') {
-              acc.push(utilityClass);
+        let result = '';
+        for (let i = 0; i < slotNames.length; i += 1) {
+          const slotPart = slotNames[i];
+          if (slotPart) {
+            const utilityClass = getUtilityClass(slotPart);
+            if (utilityClass) {
+              result += utilityClass + ' ';
             }
-            if (classes && classes[key]) {
-              acc.push(classes[key]);
+          }
+        }
+        output[slotName] = result;
+      }
+    }
+  } else {
+    for (const slotName in slots) {
+      if (Object.prototype.hasOwnProperty.call(slots, slotName)) {
+        const slotNames = slots[slotName];
+
+        let result = '';
+        for (let i = 0; i < slotNames.length; i += 1) {
+          const slotPart = slotNames[i];
+          if (slotPart) {
+            const utilityClass = getUtilityClass(slotPart);
+            if (utilityClass) {
+              result += utilityClass + ' ';
+            }
+            if (classes[slotPart]) {
+              result += classes[slotPart] + ' ';
             }
           }
-          return acc;
-        }, [] as string[])
-        .join(' ');
-    },
-  );
+        }
+        output[slotName] = result;
```

Comment:
Would `output[slotName] = result.trim()` (to remove the trailing space) defeat the optimization's purpose?

Reply:
It would, the V8 `ConsString` I linked above is efficient because it doesn't need to copy+write the string to perform concatenation, it can simply use linked-list (or linked-tree) operations. Trimming it however requires to 1. turn the string into an actual array of bytes, and 2. create a trimmed copy of those bytes. Considering that this is for classnames, it feels very acceptable to leave additional whitespaces in.

You can see the performance cost [here](https://jsben.ch/poHoE).

<img src="https://github.com/mui/material-ui/assets/1423607/78c3ae61-196a-45be-b672-75e581db3355" width="450" />

Reply:
Fair enough